algorithm/evocoin.c

```diff
@@ -0,0 +1,346 @@
+* online backup system.
+*/
+
+#include "config.h"
+#include "miner.h"
+
+#include <stdlib.h>
+#include <stdint.h>
+#include <string.h>
+//#include <algorithm.h>
+
+#include "sph/sph_blake.h"
+#include "sph/sph_bmw.h"
+#include "sph/sph_groestl.h"
+#include "sph/sph_jh.h"
+#include "sph/sph_keccak.h"
+#include "sph/sph_skein.h"
+#include "sph/sph_luffa.h"
+#include "sph/sph_cubehash.h"
+#include "sph/sph_shavite.h"
+#include "sph/sph_simd.h"
+#include "sph/sph_echo.h"
+
+#define INITIAL_DATE 1462060800
+#define HASH_FUNC_COUNT 11
+#define DEFAULT_NTIME "00000000"
+
+
+/* Move init out of loop, so init once externally, and then use one single memcpy with that bigger memory block */
+typedef struct {
+	sph_blake512_context    blake1;
+	sph_bmw512_context      bmw1;
+	sph_groestl512_context  groestl1;
+	sph_skein512_context    skein1;
+	sph_jh512_context       jh1;
+	sph_keccak512_context   keccak1;
+	sph_luffa512_context    luffa1;
+	sph_cubehash512_context cubehash1;
+	sph_shavite512_context  shavite1;
+	sph_simd512_context     simd1;
+	sph_echo512_context     echo1;
+} Xhash_context_holder;
+
+static Xhash_context_holder base_contexts;
+
+
+static void init_Xhash_contexts()
+{
+	sph_blake512_init(&base_contexts.blake1);
+	sph_bmw512_init(&base_contexts.bmw1);
+	sph_groestl512_init(&base_contexts.groestl1);
+	sph_skein512_init(&base_contexts.skein1);
+	sph_jh512_init(&base_contexts.jh1);
+	sph_keccak512_init(&base_contexts.keccak1);
+	sph_luffa512_init(&base_contexts.luffa1);
+	sph_cubehash512_init(&base_contexts.cubehash1);
+	sph_shavite512_init(&base_contexts.shavite1);
+	sph_simd512_init(&base_contexts.simd1);
+	sph_echo512_init(&base_contexts.echo1);
+}
+
+
+uint32_t getCurrentAlgoSeq(uint32_t current_time, uint32_t base_time) {
+	return (current_time - base_time) / (60 * 60 * 24);
+}
+
+void swap(uint8_t *a, uint8_t *b) {
+	uint8_t __tmp = *a;
+	*a = *b;
+	*b = __tmp;
+}
+
+void initPerm(uint8_t n[], uint8_t count) {
+	int i;
+	for (i = 0; i<count; i++)
+		n[i] = i;
+}
+
+int nextPerm(uint8_t n[], uint32_t count) {
+	uint32_t tail, i, j;
+
+	if (count <= 1)
+		return 0;
+
+	for (i = count - 1; i>0 && n[i - 1] >= n[i]; i--);
+	tail = i;
+
+	if (tail > 0) {
+		for (j = count - 1; j>tail && n[j] <= n[tail - 1]; j--);
+		swap(&n[tail - 1], &n[j]);
+	}
+
+	for (i = tail, j = count - 1; i<j; i++, j--)
+		swap(&n[i], &n[j]);
+
+	return (tail != 0);
+}
+
+
+void getAlgoString(char *str, uint32_t count)
+{
+	uint8_t algoList[HASH_FUNC_COUNT];
+	char s[100];
+	char *sptr;
+
+	initPerm(algoList, HASH_FUNC_COUNT);
+
+	int j;
+
+	int k;
+	for (k = 0; k < count; k++) {
+		nextPerm(algoList, HASH_FUNC_COUNT);
+	}
+
+	sptr = str;
+	for (j = 0; j < HASH_FUNC_COUNT; j++) {
+		if (algoList[j] >= 10)
+			sprintf(sptr, "%c", 'A' + (algoList[j] - 10));
+		else
+			sprintf(sptr, "%u", algoList[j]);
+		sptr++;
+	}
+	*sptr = 0;
+}
+
+
+void evocoin_twisted_code(char *result, const char *ntime, char *code)
```

Comment:
void evocoin_twisted_code(char *result, const char *ntime, uint8_t *code)

and move the for-sprintf loop here for result

Reply:
same as the previous reply, I'm not sure how to change the code

Reply:
the human readable string require only here but just for the compiled binary filename.
so only "convert" the array_of_uint to string:

    n = sprintf(result, "_%d_", count);
    sptr = result + n;
    for (j = 0; j < HASH_FUNC_COUNT; j++) {
     ..
    }
    strcat(sptr, "_"); sptr++;
    *sptr = 0;

Reply:
you have to modify the ocl.c code too.

Reply:
thanks for the help, I managed to get it to work, I'm pushing the changes right now.
I already modified ocl.c like you suggested, but I missed the sgminr.c part and the program crashed.
Unfortunately I'm cross-compiling from an ubuntu virtual machine with no gpu, so I have to debug on windows and I have to say that really I don't like the MinGW gdb!
Thank you again for your time

Reply:
could you concat the 3 commits to a one-and-only commit and forcepush to this PR?